C# Interceptor Support

mehrdada opened the pull request:
This PR adds core interceptor hooks to gRPC C# (both server side and client side).

Addresses https://github.com/grpc/grpc/issues/11646
Related proposal https://github.com/grpc/proposal/pull/38

jtattermusch commented:
CC @jtattermusch

mehrdada commented on src/csharp/Grpc.Core.Tests/Interceptors/ClientInterceptorTest.cs:
```diff
@@ -0,0 +1,76 @@
+            });
+            var server = helper.GetServer();
+            server.Start();
+            var callInvoker = helper.GetChannel().Intercept(interceptor);
```
Yes, that is how it is going to work.

I am considering adding another overload that takes `params Interceptor[]` and adds the list in-order (not the opposite order implied by decorator style).

Disturbing replied:
I'd definitely add this order of operations to be clear. I pref FIFO, but I would say you want both functions to behave the same versus supporting one the opposite way.

mehrdada replied:
I think FIFO does not make sense when chaining calls like `.Intercept(a).Intercept(b)`, because you can imagine a scenario in which you are given a `CallInvoker` by another part of your program that is created with `channel.Intercept` and you then decide to intercept it again and see the whole thing as a black box.

However, I do agree that an overload that takes multiple interceptors in a single call: `.Intercept(a, b, c)` should pass control to `a`, `b`, and `c` in order.

jtattermusch replied:
I think the Intercept(a, b, c) overload would be useful.

plaisted replied:
I think it'd be nice to also have an overload like:
```
.Intercept(p => {
        p.AddLogging();
        p.AddAuth();
    });
```
where p is a pipeline object of some sort.

This would make it very easy to develop packagable pipeline plugins by just creating an extension method for the pipeline object. Additionally in cases where you don't about the type of TRequest / TResponse (eg. logging) you could create a generic extension method so that your plugin will work for all types of requests / responses.

mehrdada replied:
@plaisted Can you clarify what is the advantage of that over

```csharp
.Intercept(new Logger(), new Authenticator());
```

?

plaisted replied:
@mehrdada 

Functionally they are the same but using the extension method allows you to nicely encapsulate what "API" you want the user of the package to use vs. what the actual constructor that the class needs.

As a contrived example the `Authenticator` interceptor may require a `TokenValidator` service so it's constructor is actually `Authenticator(ITokenValidator validator)`. Adding an extensions method in your package of `AddJwtAuth()` would simply construct the `Authenticator` interceptor with a `JwtTokenValidator` that implements `ITokenValidator`.

Now the extension method would generally do a lot more than that, maybe even add multiple interceptors which is trivial with this sort of setup:
```csharp
//extension method
public static void AddJwtAuth(this IInterceptorPipeline pipeline, Action<JwtOptions> optionsBuilder) {
    var options = optionsBuilder(new JwtOptions());
    pipeline.Add(new Authenticator(new JwtTokenValidator(options)));
    pipeline.AddInterceptor2(); // add a dependent interceptor
    pipeline.AddInterceptor3(); // add another
}

//using it
.Intercept(p => {
        p.AddJwtAuth(o=>{ o.ValidationCertificate = cert; });
    });
```

I've written a few addons (aspcore mainly) that use this sort of approach and it works very nicely as you don't have to expose any implementation details to the users of the package.  Admittedly all of this other than adding multiple interceptors could all be implemented relatively easily using a static "factory" method so a lot of this is preference and wanting to stay similar to how I see this done across different frameworks in c#.

mehrdada commented on src/csharp/Grpc.Core/ClientBase.cs:
```diff
@@ -147,6 +149,64 @@ internal ClientBaseConfiguration Configuration
+                /// <summary>
+                /// Intercepts an asynchronous invocation of a simple remote call.
+                /// </summary>
+                public override AsyncUnaryCall<TResponse> AsyncUnaryCall<TRequest, TResponse>(Method<TRequest, TResponse> method, string host, CallOptions options, TRequest request, Func<Method<TRequest, TResponse>, string, CallOptions, TRequest, AsyncUnaryCall<TResponse>> next)
```
How about `continuation`?

jtattermusch replied:
continuation sounds much better.

mehrdada replied:
I have changed my mind a bit on `continuation` and have reservations about it since in the retry case, you can invoke it more than once and you can choose not to call it at all. How about `invoker`, though the downside is it is not an instance of `CallInvoker`.

mehrdada commented on src/csharp/Grpc.Core/Interceptors/ClientInterceptor.cs:
```diff
@@ -0,0 +1,281 @@
+        /// <summary>
+        /// Returns a new instance of <c>AsyncServerStreamingCall</c> that has hooks to intercept on an underlying object of the same type.
+        /// </summary>
+        protected static AsyncServerStreamingCall<TResponse> Intercept<TRequest, TResponse>(AsyncServerStreamingCall<TResponse> call,
```
They are provided for interceptor implementors to be able to intercept additional events in async or streaming calls. The initial interceptor hook is only invoked at the beginning of the call.

These functions are only needed because the constructor for `AsyncXYZCall` are internal. If we are going to make them public, now would be a good time. I could just remove all of this mess.

jskeet replied:
I'd quite like a way of creating an interceptor to just modify/inspect one aspect (e.g. metadata) of all calls - `Interceptor.ForMetadata(metadata => { ...})`

mehrdada replied:
I agree, and have a few other helper extension methods in mind. I wanted to keep this PR minimal to the core hooks to make it easier to review... now I am thinking it might have been best to add such things for additional context?

Disturbing commented on src/csharp/Grpc.Core/CallOptions.cs:
```diff
@@ -36,6 +37,7 @@ public struct CallOptions
         ContextPropagationToken propagationToken;
         CallCredentials credentials;
         CallFlags flags;
+        Dictionary<object, object> items;
```
I'm a huge fan of object versus string, sometime I like to store things via by type.

Get<T>() for example.

jskeet replied:
I'm concerned that this already-large struct is becoming huge. The threading is also a concern... `Dictionary<,>` isn't thread-safe - what are the expected uses here, and are we satisfied they won't cause problems?

mehrdada replied:
@jskeet The use-case is to share some call-specific data between the interceptors and potentially the call invoker. (The closest .NET analogy I have in mind is `HttpContext.Context`.) I initially thought Metadata headers should be abused for this purpose, but it can be too inconvenient. For example, one use case is associating a `User` object (or any other state information) via an interceptor that can then be pulled and injected into the RPC.  Another design I contemplated with is replacing the dictionary with a single `object Tag`, however, from a struct size point of view, it has the same effect. It will alleviate the threading concerns though, perhaps at a cost to usability.

jtattermusch replied:
I understand the use case for "items", but I share @jskeet 's concerns.  

Idea - could we postpone adding the "items" field and later add it without breaking the API? Adding the basic functionality for interceptors in this PR and having one or two "improvements" PR might actually be productive approach.

mehrdada replied:
Happy to move `items` to a different PR, but this is an important requirement it seems, and we need to tackle it soon anyway. Metadata strings are a really ugly approach.

Another idea is to introduce an `InterceptorContext` which holds interceptor state and pass it along in the continuation chain. However, the downside is there is no clear way to share that information with the final `CallInvoker` (which might or might not be necessary). If we are going to do this, we can't do it in another PR, because that is a core API change in the interceptor itself (adding another argument to interception callbacks. and the continuation delegate. I'd really like to know what @jskeet thinks about this alternative solution.

Disturbing commented:
Another case is when u want to add inceptors dynamically in the lifecycle
of a server. (I'm considering a grpc serverless kind of deal)

May just have the interceptors optionally have sorting ability on insertion
and override compare if/when needed?
On Sat, Sep 30, 2017 at 8:18 AM Mehrdad Afshari <notifications@github.com>
wrote:

> *@mehrdada* commented on this pull request.
> ------------------------------
>
> In src/csharp/Grpc.Core.Tests/Interceptors/ClientInterceptorTest.cs
> <https://github.com/grpc/grpc/pull/12613#discussion_r141991058>:
>
> > +        const string Host = "127.0.0.1";
> +
> +        [Test]
> +        public void AddRequestHeaderInClientInterceptor()
> +        {
> +            var helper = new MockServiceHelper(Host);
> +            var interceptor = new AddHeaderClientInterceptor("x-client-interceptor", "hello world");
> +            helper.UnaryHandler = new UnaryServerMethod<string, string>((request, context) =>
> +            {
> +                var interceptorHeader = context.RequestHeaders.Last(m => (m.Key == interceptor.Header.Key)).Value;
> +                Assert.AreEqual(interceptorHeader, interceptor.Header.Value);
> +                return Task.FromResult("PASS");
> +            });
> +            var server = helper.GetServer();
> +            server.Start();
> +            var callInvoker = helper.GetChannel().Intercept(interceptor);
>
> I think FIFO does not make sense when chaining calls like
> .Intercept(a).Intercept(b), because you can imagine a scenario in which
> you are given a CallInvoker by another part of your program that is
> created with channel.Intercept and you then decide to intercept it again
> and see the whole thing as a black box.
>
> However, I do agree that an overload that takes multiple interceptors in a
> single call: .Intercept(a, b, c) should pass control to a, b, and c in
> order.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/grpc/grpc/pull/12613#discussion_r141991058>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ACQev48NFOUj_UUUCOrLLjQ9JK8ytdHYks5snYjngaJpZM4PavYc>
> .
>

jskeet commented on src/csharp/Grpc.Core.Tests/MockServiceHelper.cs:
```diff
@@ -184,6 +184,10 @@ public ServerServiceDefinition ServiceDefinition
             {
                 return this.serviceDefinition;
             }
+            set
```
Just turn this into an automatically implemented property? (Ditto for Host potentially, but that's not being changed in this PR, so it may want to be done separately.)

mehrdada replied:
I considered this, but all (or almost all?) properties in gRPC C# code base seem to be not automatically generated. I wonder if it is intentional for compatibility purposes with older compilers. That's why I kept consistent with the approach in the codebase.

jtattermusch replied:
ad auto-properties: there's no particular reason why we are not using them, but you're right the most times I didn't use them so for now it might makes sense to stay consistent.

I think autoproperties have been in C# since early version, so compiler compatibility is not a concern here. But there are other language constructs (from C# 6) that we intentionally didn't use because until about a year ago, we needed the code to compile on older mono and Visual Studio. Now when we switched to the new dotnet projects, we could start using C# 6 in our code (the compiled code will still be compatible with older .NET and the generated code needs to still compile with older .NET) - so feel free to use C# 6 if you want to. Sometime in the future, we might refactor portions of our existing code and introduce C# 6 constructs where appropriate.

jskeet commented on src/csharp/Grpc.Core/CallOptions.cs:
```diff
@@ -203,6 +216,17 @@ public CallOptions WithCredentials(CallCredentials credentials)
 
         /// <summary>
+        /// Returns a new instance of <see cref="CallOptions"/> with <c>Items</c> set to the value provided.
+        /// Values of all other fields are preserved.
```
Possibly clarify that any *previous* items are effectively lost rather than merged?

jskeet commented on src/csharp/Grpc.Core/Interceptors/ServerInterceptor.cs:
```diff
@@ -0,0 +1,150 @@
+            }
+        }
+
+        private Delegate WrapDelegate(Delegate d)
```
All this reflection looks inefficient to me. I haven't quite got my head round how it's all used at the moment, but it would be great to avoid it...

mehrdada replied:
This process only happens at the startup time of the service once, so I don't think the potential performance impact matters. Unfortunately, I don't know of any other way to avoid this dance, since `dynamic` or using C# binder to find the appropriate overload require depending on additional libraries. We need this to map a non-generic object back to the correct parameterized-type overload. If you have a better way to accomplish this in mind, definitely let me know please.

mehrdada commented on src/csharp/Grpc.Core/Interceptors/ClientInterceptor.cs:
```diff
@@ -0,0 +1,281 @@
+            if (responseHeaders != null)
+            {
+                var callResponseHeadersOriginal = callResponseHeaders; // Ensure it will not be captured by the closure
+                callResponseHeaders = new Task<Metadata>(() => responseHeaders(callResponseHeadersOriginal));
```
I have thought about this a bit more and I think there is no good reason for this set of functions to exist, whose purpose is to circumvent the `internal`-ness of the constructors of `Async***Call` objects. If we open up the constructors, this whole thing becomes unnecessary.

jtattermusch replied:
Downsides:
we'd need to polish the Async**Call constructor signature and be certain that it's not going to change (that would be an API breakage and we are trying hard to avoid that). At the same we would lose some ability to add new members to the Async**Call (what if you need to add more arguments to the constructor later?).
Also, more generally, there is always some risk associated with exposing lower-level APIs to end users, because non-expert users tend to shoot themselves in the foot if you expose too much of the internals - and gRPC users shouldn't be required to become experts before they can use gRPC efficiently.

That said, I think it makes sense to revisit making Async**Call constructors public - now that we have an official support for interceptors coming.

FTR, here's an external PR for publishing the Async**Call constructor - but I think a bit more refactoring would be useful https://github.com/grpc/grpc/pull/9330/files.

mehrdada replied:
> we'd need to polish the `Async*Call` constructor signature and be certain that it's not going to change

The members are already public, so we cannot remove them and the constructor at the moment is basically initializing the existing members. We will be able to add additional constructors should we add new members (as long as we can find sensible defaults for values). In effect, by exposing this functionality within this PR under the interceptor API, we are transitively doing most of this API retention cost (admittedly it is a bit contained, but still would be breaking). I think we should seriously consider opening them up, since the class itself and its members are already not `internal` and the `InterceptAsync*` methods are really ugly in this PR.

I do agree that polishing the constructor now makes sense, but I don't see anything wrong with it. What would be the ideal in your opinion?

jtattermusch replied:
Yes, let's look into publishing the Async**Call constructors in a separate PR (and count on them being already public in this PR).

mehrdada replied:
What do you think about not opening the constructors but adding a static factory method that returns a new instance? The advantage would be that we will retain the freedom to *unseal* the class and return a derived instance without an API breakage, but it is debatable that it would be worth much.

mehrdada replied:
The PR for publicizing constructors is here: https://github.com/grpc/grpc/pull/12877 though I am still open to the static factory route idea.

jtattermusch commented:
As @jskeet suggests, I think we really should implement several interceptors to get the feel of the API, that basically goes hand in hand with my request for snippets that demonstrate common use cases - once we have those implementations/snippets, we'll know more.

mehrdada commented:
@Disturbing I am not sure there is a good way for the core API to be able to handle dynamic insertion and removal without incurring API complications/potential performance costs. However, a simple solution for the user, given the current API, would be to write a "MultiplexerInterceptor" that is a meta-interceptor and supports insertion and removal if they really need that feature and are willing to pay the cost.

Disturbing commented:
Agreed.

Although the cost is minimal i would imagine with a quick lookup in some
sort of header or method name hash registry. Def not looking to proxy or
use reflection like I am now and awaiting this inceptor implementation.
On Fri, Oct 6, 2017 at 2:13 AM Mehrdad Afshari <notifications@github.com>
wrote:

> @Disturbing <https://github.com/disturbing> I am not sure there is a good
> way for the core API to be able to handle dynamic insertion and removal
> without incurring API complications/potential performance costs. However, a
> simple solution for the user, given the current API, would be to write a
> "MultiplexerInterceptor" that is a meta-interceptor and supports insertion
> and removal if they really need that feature and are willing to pay the
> cost.
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/grpc/grpc/pull/12613#issuecomment-334547670>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ACQev4QlRZ03h80plEMjKJq3K_9vanXmks5spRw5gaJpZM4PavYc>
> .
>

mehrdada commented:
@Disturbing I mostly worry about threading issues and the need to add access the dictionary concurrently, adding locks, and possibly slowing down all requests even if you don't intend to do any add/remove from the registry. Also, defining the semantics to make it clear what would happen when you remove an interceptor for an in-flight request, for example, can make things complicated and potentially hard to stick to across future versions.

Disturbing commented:
True, i would go with a read write lock over a concurrent dictionary though
due to how writes would be very rare.

To explain my use case a bit more. I'm imagining making a serverless
hotswappable grpc environment which would require this kind of
functionality for the particular microservice loaded. :)

I would build this on top of what gets in as the core interceptors for
client and server is a must have and the dynamic part is a nice to have.

On Fri, Oct 6, 2017 at 11:32 AM Mehrdad Afshari <notifications@github.com>
wrote:

> @Disturbing <https://github.com/disturbing> I mostly worry about
> threading issues and the need to add access the dictionary concurrently,
> adding locks, and possibly slowing down all requests even if you don't
> intend to do any add/remove from the registry. Also, defining the semantics
> to make it clear what would happen when you remove an interceptor for an
> in-flight request, for example, can make things complicated and potentially
> hard to stick to across future versions.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/grpc/grpc/pull/12613#issuecomment-334649323>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ACQev3AsN3ttceKfLWeVK8RMQylWMf4Eks5spZ9PgaJpZM4PavYc>
> .
>

Disturbing commented:
Reread your comment. Realized you mentioned locks as the solution vs
concurrent dictionary. Rwlslim will add ~50 nanoseconds to the call and
having the option to enable or disable such functionality by a null check
otherwise a separate class implementation like the multiplexerinterceptor
is fine.


On Fri, Oct 6, 2017 at 11:47 AM Joseph Cooper <jdcooper4@gmail.com> wrote:

> True, i would go with a read write lock over a concurrent dictionary
> though due to how writes would be very rare.
>
> To explain my use case a bit more. I'm imagining making a serverless
> hotswappable grpc environment which would require this kind of
> functionality for the particular microservice loaded. :)
>
> I would build this on top of what gets in as the core interceptors for
> client and server is a must have and the dynamic part is a nice to have.
>
> On Fri, Oct 6, 2017 at 11:32 AM Mehrdad Afshari <notifications@github.com>
> wrote:
>
>> @Disturbing <https://github.com/disturbing> I mostly worry about
>> threading issues and the need to add access the dictionary concurrently,
>> adding locks, and possibly slowing down all requests even if you don't
>> intend to do any add/remove from the registry. Also, defining the semantics
>> to make it clear what would happen when you remove an interceptor for an
>> in-flight request, for example, can make things complicated and potentially
>> hard to stick to across future versions.
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/grpc/grpc/pull/12613#issuecomment-334649323>, or mute
>> the thread
>> <https://github.com/notifications/unsubscribe-auth/ACQev3AsN3ttceKfLWeVK8RMQylWMf4Eks5spZ9PgaJpZM4PavYc>
>> .
>>
>

mehrdada commented:
@Disturbing Yeah, by "locking" I meant it more generally than the `lock` statement. I am most worried about the additional complexity.  Let me think about it a bit more, but I don't think a mere dictionary alone will cut it, since I care about order of interceptors being invoked. I think if we end up doing it, it will effectively look like that `MultiplexerInterceptor` I just described, but shipped in the `Grpc.Core` assembly. I'll focus on the core functionality in this PR; we can add that later.

chwarr commented on src/csharp/Grpc.Core.Tests/Interceptors/ServerInterceptorTest.cs:
```diff
@@ -0,0 +1,78 @@
+            {
+                this.header = new Metadata.Entry(key, value);
+            }
+            public override async Task<TResponse> UnaryServerHandler<TRequest, TResponse>(TRequest request, ServerCallContext context, UnaryServerMethod<TRequest, TResponse> next)
```
This method is async and `return await next()`, while the client test is non-async-but-still-Task and just returns `next()`. Is the difference intentional?

mehrdada replied:
Which client method? Line 69?

chwarr replied:
I was referring to difference between the client interceprot test 
 (in [`ClientInterceptorTest.AddHeaderClientInterceptor.BlockingUnaryCall`](https://github.com/grpc/grpc/pull/12613/files#diff-000199ab016d959fb86fd61611223f6eR42)) and the server interceptor test.

The relevant client test code is:
```
public override TResponse BlockingUnaryCall<TRequest, TResponse>(Method<TRequest, TResponse> method, string host, CallOptions options, TRequest request, Func<Method<TRequest, TResponse>, string, CallOptions, TRequest, TResponse> next)
{
    options.Headers.Add(this.header);
    return next(method, host, options, request);
}
```

It just returns a Task, while in the similar `ServerInterceptorTest.AddRequestHeaderServerInterceptor.UnaryServerHandler`, the method is marked `async` and has the pattern `return await next(...)`. Since there's no other async action in this method, I would have expected a non-async method that just `return next(...);`

Is there something special about the server-side or is this an unintentional difference?

mehrdada replied:
Thanks for pointing it out. Fixed.

JordyLangen commented:
Any reason this API differs from the Java implementation? 

[ClientInterceptor](https://github.com/grpc/grpc-java/blob/166108a9438c22d06eb3b371b5ad34a75e14787c/core/src/main/java/io/grpc/ClientInterceptor.java)

[ClientInterceptors](https://github.com/grpc/grpc-java/blob/166108a9438c22d06eb3b371b5ad34a75e14787c/core/src/main/java/io/grpc/ClientInterceptors.java)


[ServerInterceptor](https://github.com/grpc/grpc-java/blob/166108a9438c22d06eb3b371b5ad34a75e14787c/core/src/main/java/io/grpc/ServerInterceptor.java)

[ServerInterceptors](https://github.com/grpc/grpc-java/blob/166108a9438c22d06eb3b371b5ad34a75e14787c/core/src/main/java/io/grpc/ServerInterceptors.java)

Just wondering because I see some added value for common APIs between language implementations coming from an environment that uses multiple languages/platforms.

mehrdada commented:
@JordyLangen There's value in consistency across implementations, and C# implementation is arguably more similar to Go, Ruby, and upcoming Python designs, leaving Java as the odd duck here. I think the differences in existing APIs are partially the cause for differences. I would say the updated C# server interceptors are more similar to Java style. On the client side, Java's implementation is very much tied to the `Channel`, whereas in C# we had the nice `CallInvoker` abstraction already existing. One design goal was to interfere in the core RPC machinery as little as possible, so as to (1) not potentially hinder the performance of the non-intercepted RPCs. (2) be able to short-circuit the whole thing. (I also think part of the root cause for some things can be traced back to differences in generics in C# and Java.)
This design (C#), basically gives control to the interceptor to do anything it wants right when a call is made. It is easy to build an interceptor on top of this that raises event notifications as headers get sent, messages get sent, etc. more similar to the Java ones.

vassilvk commented:
@mehrdada - what is the status of this PR? Was it replaced by another implementation?

mehrdada commented:
@vassilvk I am planning to get back to this as soon as we get done with the imminent gRPC 1.8 release. I am considering some potential design tweaks though. Stay tuned.

HideDev commented:
Hi, any news?

mehrdada commented:
@HideDev actively polishing some final stuff. will have an update for you this week

HideDev commented:
@mehrdada thanks

893949088 commented:
Hi,Is there no new progress?

mehrdada commented:
Cleaned-up, polished, and rebased. Also added `GenericInterceptor` which massively simplifies authorship of interceptors. Added a couple more example interceptors embedded in tests, etc.

@jtattermusch PTAL and let's get the ball rolling quickly. thanks!

jtattermusch commented on src/csharp/Grpc.Core/Interceptors/CallInvokerExtensions.cs:
```diff
@@ -0,0 +1,171 @@
+        /// <param name="invoker">The channel to intercept.</param>
+        /// <param name="interceptors">
+        /// An array of interceptors to intercept the calls to the invoker with.
+        /// Control is passed to the interceptors in the order specified.
```
perhaps also add a remark that  the order of interceptors in  the array is  the reverse of what you get when you use passing via .Intercept(a).Intercept(b).Intercept(c)

mehrdada replied:
Done

jtattermusch commented on src/csharp/Grpc.Core/Interceptors/CallInvokerExtensions.cs:
```diff
@@ -0,0 +1,171 @@
+        /// </param>
+        public static CallInvoker Intercept(this CallInvoker invoker, params Interceptor[] interceptors)
+        {
+            if (interceptors == null)
```
GrpcPreconditions.CheckNotNull() instead of silently accepting null.

jtattermusch replied:
also add tests that null  is not accepted as `interceptor` or `interceptors` here or above.

jtattermusch commented on src/csharp/Grpc.Core/Interceptors/CallInvokerExtensions.cs:
```diff
@@ -0,0 +1,171 @@
+        /// An array of interceptors to intercept the calls to the invoker with.
+        /// Control is passed to the interceptors in the order specified.
+        /// </param>
+        public static CallInvoker Intercept(this CallInvoker invoker, params Interceptor[] interceptors)
```
I think this overload is untested. Please add a test if so.

jtattermusch commented on src/csharp/Grpc.Core/Interceptors/ChannelExtensions.cs:
```diff
@@ -0,0 +1,65 @@
+    /// Provides extension methods to make it easy to register interceptors on Channel objects.
+    /// This is an EXPERIMENTAL API.
+    /// </summary>
+    public static class ChannelExtensions
```
something to consider: 
if these extensions are under Grpc.Core.Interceptors namespace, then the extension methods won't be visible to code completion unless you  add `using Grpc.Core.Interceptor;`  to the source. Not sure if that's desired or not.

mehrdada replied:
That is intended indeed, but let me know what is preferable to you. I am 60-40 in favor of the current organization.

jtattermusch commented on src/csharp/Grpc.Core/Interceptors/GenericInterceptor.cs:
```diff
@@ -0,0 +1,449 @@
+        /// <summary>
+        /// Provides hooks through which an invocation should be intercepted.
+        /// </summary>
+        public sealed class ClientCallArbitrator<TRequest, TResponse>
```
I'm not super keen on the name "arbitrator", but I have to give it more thought (Are there any alternative names to consider?).

mehrdada replied:
How about `Controller`? I also considered `Observer` and `Listener`, but there are properties that let you override and control as it goes by, so it is not a passive object, necessarily.

mehrdada commented on src/csharp/Grpc.Core.Tests/Interceptors/ClientInterceptorTest.cs:
```diff
@@ -0,0 +1,132 @@
+                metadata.Add(new Metadata.Entry(HeaderKey, HeaderValue));
+                return metadata;
+            });
+            Assert.AreEqual("PASS", callInvoker.BlockingUnaryCall(new Method<string, string>(MethodType.Unary, MockServiceHelper.ServiceName, "Unary", Marshallers.StringMarshaller, Marshallers.StringMarshaller), Host, new CallOptions().WithHeaders(new Metadata()), ""));
```
Done

mehrdada commented on src/csharp/Grpc.Core.Tests/Interceptors/ClientInterceptorTest.cs:
```diff
@@ -0,0 +1,132 @@
+                return new ClientCallArbitrator<TRequest, TResponse>
+                {
+                    OnRequestMessage = m => { counter++; return m; },
+                    OnUnaryResponse = x => (TResponse)(object)counter.ToString()
```
Done

mehrdada commented on src/csharp/Grpc.Core.Tests/Interceptors/ServerInterceptorTest.cs:
```diff
@@ -0,0 +1,113 @@
+            {
+                return Task.FromResult("PASS");
+            });
+            string s = "";
```
Done

mehrdada commented on src/csharp/Grpc.Core/Internal/ServerCallHandler.cs:
```diff
@@ -147,9 +163,20 @@ public ServerStreamingServerCallHandler(Method<TRequest, TResponse> method, Serv
             }
             await finishedTask.ConfigureAwait(false);
         }
+
```
Done

mehrdada commented on src/csharp/Grpc.Core/Internal/ServerCallHandler.cs:
```diff
@@ -147,9 +163,20 @@ public ServerStreamingServerCallHandler(Method<TRequest, TResponse> method, Serv
+
+        IServerCallHandler IInterceptableCallHandler.Intercept(Interceptor interceptor)
+        {
+            if (interceptor == null)
```
This is a method that only can be called by internal code. The surface of the interceptor API should check for null before ever invoking this.  I will remove this check entirely.

mehrdada commented on src/csharp/Grpc.Core/ClientBase.cs:
```diff
@@ -158,7 +182,7 @@ internal ClientBaseConfiguration(CallInvoker undecoratedCallInvoker, string host
             internal CallInvoker CreateDecoratedCallInvoker()
             {
-                return new InterceptingCallInvoker(undecoratedCallInvoker, hostInterceptor: (h) => host);
+                return undecoratedCallInvoker.Intercept(new ClientHeaderInterceptor((method, host, options) => Tuple.Create(this.host, options)));
```
This can be refactored but I leave more changes here to a separate PR, as it might propagate and result in unrelated changes, so I will leave this change minimal.

mehrdada commented on src/csharp/Grpc.Core/ClientBase.cs:
```diff
@@ -147,6 +149,28 @@ internal ClientBaseConfiguration Configuration
+
+                protected override ClientCallArbitrator<TRequest, TResponse> InterceptCall<TRequest, TResponse>(ClientInterceptorContext<TRequest, TResponse> context, bool clientStreaming, bool serverStreaming, TRequest request)
+                {
+                    var newHeaders = interceptor(context.Method, context.Host, context.Options);
```
Done

mehrdada commented on src/csharp/Grpc.Core/Interceptors/CallInvokerExtensions.cs:
```diff
@@ -0,0 +1,171 @@
+        /// </summary>
+        /// <param name="invoker">The underlying invoker to intercept.</param>
+        /// <param name="interceptor">The metadata interceptor to intercept calls to the invoker with.</param>
+        public static CallInvoker Intercept(this CallInvoker invoker, Func<Metadata, Metadata> interceptor)
```
Done

jtattermusch commented on src/csharp/Grpc.Core/ServerServiceDefinition.cs:
```diff
@@ -45,6 +48,28 @@ private ServerServiceDefinition(Dictionary<string, IServerCallHandler> callHandl
+        /// This is an EXPERIMENTAL API.
+        /// </summary>
+        /// <param name="interceptor">The interceptor to register on service.</param>
+        public ServerServiceDefinition Intercept(Interceptor interceptor)
```
a general thought on the Intercept(interceptor) methods:
there's a distinction between  the act of registering an interceptor    and   actually intercepting the call. In this PR, we are using  "Intercept" in the sense of "register interceptor", but the name "intercept" might have the feel of actually intercepting something.  Would renaming "InterceptWith(interceptor)" be clearer for the users? What does the Java API do here?
I dont' feel strongly about this, just wanted discuss the idea a bit.

mehrdada replied:
Java has `withInterceptor`, but I think it is pretty clear and more concise and C#y as is. The type of the receiver class should make it clear to the user what the operation being done is (i.e. Channel).

jtattermusch commented on src/csharp/Grpc.Core/ServerServiceDefinition.cs:
```diff
@@ -45,6 +48,28 @@ private ServerServiceDefinition(Dictionary<string, IServerCallHandler> callHandl
+                x => x.Key, x =>
+                {
+                    var value = x.Value;
+                    var interceptable = value as IInterceptableCallHandler;
```
maybe I don't understand correctly, but it looks a bit like "manual" dispatch (in the sense of not utilizing polymorphism correctly). 

In what situation won't the value be not an instance of IInterceptableCallHandler?

mehrdada replied:
<strike>We could have fused the two interfaces together (i.e. `IInterceptableCallHandler` into `IServerCallHandler` and all of them are internal so it doesn't really make any difference. I think in particular I had the `Unimplemented` implementation in mind, but we can simply add a no-op method to that and merge the interfaces. </strike>

Merged it into the `IServerCallHandler` interface.

jtattermusch commented on src/csharp/Grpc.Core/Interceptors/CallInvokerExtensions.cs:
```diff
@@ -0,0 +1,171 @@
+            {
+                return interceptor.BlockingUnaryCall(
+                    request,
+                    new ClientInterceptorContext<TRequest, TResponse>(method, host, options),
```
btw, for high througput scenarios, allocating a new context  and a new lambda `(req, ctx) =>`  might be a significant performance hit as  extra allocations put additional load on GC (which seems to be the bottleneck according to my [not fully conclusive] observations). Not sure if we can do anything about this, here, so basically just FYI.

You also might consider making ClientInterceptorContext a struct (but that might have other challenges).

mehrdada replied:
Thinking about it, making it a struct is probably a better idea. It should be small enough and it is easier to avoid perf issues with the struct than a class down the road.

mehrdada commented on src/csharp/Grpc.Core/Interceptors/GenericInterceptor.cs:
```diff
@@ -0,0 +1,449 @@
+                this.onStreamEnd = onStreamEnd;
+            }
+
+            public void Dispose() => ((IDisposable)reader).Dispose();
```
I thought about that but there is not much we can do about the former. 

About the latter, I thought about it and am quite ambivalent about struct vs sealed class here. There are pros and cons. I will go with struct if your hunch is that'd be better.

mehrdada commented on src/csharp/Grpc.Core/Interceptors/GenericInterceptor.cs:
```diff
@@ -0,0 +1,449 @@
+    /// Provides a base class for generic interceptor implementations that raises
+    /// events and hooks to control the RPC lifecycle.
+    /// </summary>
+    public abstract class GenericInterceptor : Interceptor
```
We could have fused the two interfaces together (i.e. `IInterceptableCallHandler` into `IServerCallHandler` and all of them are internal so it doesn't really make any difference. I think in particular I had the `Unimplemented` implementation in mind, but we can simply add a no-op method to that and merge the interfaces.

mehrdada replied:
Yes, but learning from Python, I feel like a standard `GenericInterceptor` being part of our library is valuable, as without it, most interceptors will be much more verbose to write. We can move it to the tests or examples now and maybe add it in a future implementation, but that will possibly fragment other interceptor libraries that will be written on top of it (which is probably fine). 

(for some reason Github shifted my comments between posts!)

mehrdada replied:
I have turned `GenericInterceptor` into an internal class for now, as it simplifies our own internal interceptor uses (for Metadata and in ClientBase), but it is good to leave it out for now as it will be a major API-stability commitment.

mehrdada commented on src/csharp/Grpc.Core/Interceptors/Interceptor.cs:
```diff
@@ -0,0 +1,281 @@
+        /// The <see cref="Grpc.Core.Interceptors.ClientInterceptorContext{TRequest, TResponse}"/>
+        /// instance to pass to the next step in the invocation process.
+        /// </param>
+        public delegate AsyncUnaryCall<TResponse> AsyncUnaryCallContinuation<TRequest, TResponse>(TRequest request, ClientInterceptorContext<TRequest, TResponse> context)
```
Yes, it won't compile without it.

mehrdada commented on src/csharp/Grpc.Core/Interceptors/Interceptor.cs:
```diff
@@ -0,0 +1,281 @@
+        /// The callback that continues the invocation process.
+        /// This can be invoked zero or more times by the interceptor.
+        /// </param>
+        /// <returns>The response messaage of the current invocation.</returns>
```
Yes, but learning from Python, I feel like a standard `GenericInterceptor` being part of our library is valuable, as without it, most interceptors will be much more verbose to write. We can move it to the examples directory now and maybe add it in a future implementation, but that will possibly fragment other interceptor libraries that will be written on top of it (which is probably fine). Will fix the tests.

mehrdada replied:
Done

mehrdada commented on src/csharp/Grpc.Core/Interceptors/Interceptor.cs:
```diff
@@ -0,0 +1,281 @@
+        /// The <see cref="Grpc.Core.Interceptors.ClientInterceptorContext{TRequest, TResponse}"/>
+        /// instance to pass to the next step in the invocation process.
+        /// </param>
+        public delegate TResponse BlockingUnaryCallContinuation<TRequest, TResponse>(TRequest request, ClientInterceptorContext<TRequest, TResponse> context)
```
Yes, I think so. It is only really used by the derived classes of GenericInterceptor and therefore always in scope when you derive a class.

mehrdada commented on src/csharp/Grpc.Core/Interceptors/Interceptor.cs:
```diff
@@ -0,0 +1,281 @@
+    /// Carries along the context associated with intercepted invocations on the client side.
+    /// This is an EXPERIMENTAL API.
+    /// </summary>
+    public class ClientInterceptorContext<TRequest, TResponse>
```
Yeah, I'll move them in. (In a prior experiment, they were used outside as well, and that's why they got out in the first place)

mehrdada commented on src/csharp/Grpc.Core/Interceptors/GenericInterceptor.cs:
```diff
@@ -0,0 +1,449 @@
+        /// <summary>
+        /// Provides hooks through which a server-side handler should be intercepted.
+        /// </summary>
+        public sealed class ServerCallArbitrator<TRequest, TResponse>
```
I think so, as they are only usable there. It's now internal so we can change it later should we want to.

mehrdada commented on src/csharp/Grpc.Core/Interceptors/Interceptor.cs:
```diff
@@ -0,0 +1,406 @@
+    /// Serves as the base class for gRPC interceptors.
+    /// This is an EXPERIMENTAL API.
+    /// </summary>
+    public abstract class Interceptor
```
Because it is an abstract class rather than interface, I deliberately decided to merge the two interfaces into one. The overhead is minimal and it will allow one class to implement both sides, which is nice, if you are a library author for example, and provide a tracing or logging interceptor, you expose one class to the user that they register on both sides. If you want, you can implement only one side, but if we went with two base classes, there was no way to consolidate them into one implementation.

mehrdada commented:
@jtattermusch Eliminated GenericInterceptor. PTAL.

mehrdada commented:
Test failures:
Artifact build windows: shows pending on github, but log shows success
Multilang Linux: Infra error -- ignored (local csharp tests pass)

893949088 commented:
Awesome，When will this version be released？

zsojma commented:
Is there some manual how to use it? Thanks

Falco20019 commented:
You can have a look at my GRPC-Opentracing implementation: https://github.com/Falco20019/grpc-opentracing/tree/csharp-experimental/csharp/GrpcOpenTracing

You can implement your interception logic by extending the `Interceptor` class: [TracingInterceptor](https://github.com/Falco20019/grpc-opentracing/blob/csharp-experimental/csharp/GrpcOpenTracing/GrpcOpenTracing/TracingInterceptor.cs) You can do whatever you want in there and just have to call and return the continuation delegate with the (maybe modified) data.

On the usage side, you can intercept an `Channel` to get an `CallInvoker` that you can use with the generated clients (see [here](https://github.com/Falco20019/grpc-opentracing/blob/csharp-experimental/csharp/GrpcOpenTracing/GrpcOpenTracingTest/Program.cs#L45)). On the server side, you intercept the service (see [here](https://github.com/Falco20019/grpc-opentracing/blob/csharp-experimental/csharp/GrpcOpenTracing/GrpcOpenTracingTest/Program.cs#L42)).

kkolstad commented:
For those trying to follow the links in the previous comment:
* https://github.com/Falco20019/grpc-opentracing/tree/csharp-experimental/csharp/Grpc.OpenTracing
* [TracingInterceptor](https://github.com/Falco20019/grpc-opentracing/blob/csharp-experimental/csharp/Grpc.OpenTracing/TracingInterceptor.cs)
* [Usage links](https://github.com/Falco20019/grpc-opentracing/blob/csharp-experimental/csharp/Grpc.OpenTracing.Test/Program.cs)

Falco20019 commented:
@kkolstad Thanks for noticing. I updated all the links in my post so that it’s easier to follow along again :)
